src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1824,15 +1824,23 @@ private void HandleHttp2ConnectionFailure(HttpConnectionWaiter<Http2Connection?>
         /// Called when an HttpConnection from this pool is no longer usable.
         /// Note, this is always called from HttpConnection.Dispose, which is a bit different than how HTTP2 works.
         /// </summary>
-        public void InvalidateHttp11Connection(HttpConnection connection, bool disposing = true)
+        public void InvalidateHttp11Connection(HttpConnection connection, bool fromReadAheadTask = false)
         {
-            lock (SyncObj)
+            // Not locking on SyncObj here as it's expected that we may already be holding the lock in
+            // case we just added an idle connection to the pool and called OnAddingIdleConnectionToPool,
+            // which started a read-ahead that immediately completed, which in turn Disposed and Invalidated
+            // the connection.
```

Comment:
Should we instead just remove the assert from SyncObj and lock on SyncObj here as is the intent?